---
title: "Lun2"
author: "DUO Hongrui"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Simulation method: Lun2}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---
<!-- github markdown built using 
rmarkdown::render("vignettes/Lun2.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Here Lun2 method will be demonstrated clearly and hope that this document can help you.

# Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
dim(ref_data)
```

Using `simmethods::Lun2_estimation` command to execute the estimation step.

**Note: Lun2 needs cell group information or plates as input to estimate parameters**

In simmethods, cell group information of the reference data already exits. We should use this information by typing `other_prior = list(group.condition = xxx)`.
```{r estimate}
group_condition <- as.numeric(simmethods::group_condition)
estimate_result <- simmethods::Lun2_estimation(ref_data = ref_data,
                                               other_prior = list(group.condition = group_condition),
                                               verbose = T,
                                               seed = 10)
```

# Simulating datasets using Lun2

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Determin the number of cells and genes
3. Simulate two groups

## Datasets with default parameters

The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group of cells.

```{r default_params}
simulate_result <- simmethods::Lun2_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
head(colData(SCE_result))
head(rowData(SCE_result))
```

## Determin the number of cells and genes

In Lun2, we can not set `nCells` directly and should set `cell.plates` instead. For example, if we want to simulate 500 cells with one group, we must specify which plate or group each cell belongs to, e.g. `other_prior = list(cell.plates = rep(1, 500))`. For genes, we can just set `nGenes`.
Here, we simulate a new dataset with 500 cells and 1000 genes:
```{r customed cell and gene number}
simulate_result <- simmethods::Lun2_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(cell.plates = rep(1, 500),
                     nGenes = 1000),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```

## Simulate two or more groups

In Lun2, we can not set `nGroups` directly and the number of groups depends on the `cell.plates` parameter. For example, if specified cells are from three groups or plates by `cell.plates = sample(1:3, 500, replace = TRUE)` and the result will contain three groups.

For demonstration, we will simulate three groups using the learned parameters.
```{r groups}
simulate_result <- simmethods::Lun2_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(cell.plates = c(rep(1, 100),
                                     rep(2, 100),
                                     rep(3, 100)),
                     nGenes = 1000),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$plate)
```
We can alse specify the proportion of DEGs and fold change value between the groups of cells.
```{r fold change}
simulate_result <- simmethods::Lun2_simulation(
  parameters = estimate_result[["estimate_result"]],
  other_prior = list(cell.plates = sample(1:2, 500, replace = TRUE),
                     nGenes = 5000,
                     de.prob = 0.2,
                     fc.group = 2),
  return_format = "list",
  verbose = TRUE,
  seed = 111
)
count_data <- simulate_result[["simulate_result"]][["count_data"]]
dim(count_data)
## cell information
col_data <- simulate_result[["simulate_result"]][["col_meta"]]
head(col_data)
table(col_data$plate)
## gene information
row_data <- simulate_result[["simulate_result"]][["row_meta"]]
head(row_data)
### The result of Lun2 contains the factors of different groups and uses can
### calculate the fold change by division. For example, the DEFactors of A gene
### in Ingroup and Outgroup are respectively 2 and 1, and the fold change of A gene
### is 2/1=2 or 1/2=0.5.
fc_ingroup_to_outgroup <- row_data$DEFacIngroup/row_data$DEFacOutgroup
table(fc_ingroup_to_outgroup != 1)[2]/5000 ## de.prob = 0.2
### number of all DEGs
table(row_data$de_gene)[2]/5000 ## de.prob = 0.2
```
