---
title: "SCRIP"
author: "DUO Hongrui"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Simulation method: SCRIP}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---
<!-- github markdown built using 
rmarkdown::render("vignettes/SCRIP.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Here SCRIP method will be demonstrated clearly and hope that this document can help you.

# Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
dim(ref_data)
```

Using `simmethods::SCRIP_estimation` command to execute the estimation step.

```{r estimate}
estimate_result <- simmethods::SCRIP_estimation(ref_data = ref_data,
                                                verbose = T,
                                                seed = 10)
```

# Simulating datasets using SCRIP

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Determin the number of cells and genes
3. Simulate two or more groups
4. Simulate two or more batches
5. Simulate more groups and batches simutaniously
6. Simulate with five different modes

## Datasets with default parameters

The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group and one batch of cells.

**Note: SCRIP needs reference data when simulating new datasets.**
```{r default_params}
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
head(colData(SCE_result))
head(rowData(SCE_result))
```

## Determin the number of cells and genes

In SCRIP, we can not set `nCells` directly and should set `batchCells` instead. For example, if we want to simulate 500 cells, we can type `other_prior = list(batchCells = 500)`. For genes, we can just set `nGenes`.
Here, we simulate a new dataset with 500 cells and 1000 genes:
```{r customed cell and gene number}
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(batchCells = 500,
                     nGenes = 1000),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```

## Simulate two or more groups

In SCRIP, we can not set `nGroups` directly and should set `prob.group` instead. For example, if we want to simulate 2 groups, we can type `other_prior = list(prob.group = c(0.5, 0.5))`. Note that the sum of `prob.group` numeric vector must equal to 1, so we can also set `prob.group = c(0.3, 0.7)`.

In addtion, if we want to simulate three or more groups, we should obey the rules:

* The length of `prob.group` vector must always equal to the number of groups.
* The sum of `prob.group` numeric vector must equal to 1.

For demonstration, we will simulate three groups using the learned parameters.
```{r groups}
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(batchCells = 500,
                     nGenes = 1000,
                     prob.group = c(0.1, 0.3, 0.6)),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$group)

## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.1
```
We can see that the proportion of differential expressed genes is 0.1 (equals to the default). Next, if we want to know the fold change between two groups, we will do division with the groups that we are interested in.
```{r fold change}
## fc between group2 and group1
fc_group1_to_group2 <- gene_info$DEFacGroup2/gene_info$DEFacGroup1
## fc between group3 and group1
fc_group1_to_group3 <- gene_info$DEFacGroup3/gene_info$DEFacGroup1
## fc between group3 and group2
fc_group2_to_group3 <- gene_info$DEFacGroup3/gene_info$DEFacGroup2
```


## Simulate two or more batches

In SCRIP, we can not set `nBatches` directly and should set `batchCells` instead. For example, if we want to simulate 2 batches, we can type `other_prior = list(batchCells = c(250, 250))`. Note that the sum of `batchCells` numeric vector represents the total number of cells and the length of the vector equals to the number of batches.

In addtion, if we want to simulate three or more batches, we should obey the rules:

* The length of `batchCells` vector always equals to the number of batches.
* The sum of `batchCells` numeric vector represents the total number of cells.

For demonstration, we will simulate three batches using the learned parameters.
```{r batches}
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(batchCells = c(200, 300),
                     nGenes = 1000),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$batch)
```


## Simulate more groups and batches simutaniously

As mentioned before, we can set `prob.group` and `batchCells` to determine the number of groups and batches and we can also set `de.prob` to specify the proportion of DEGs. Here, we simulate a dataset with following settings:

* 1000 cells
* 5000 genes
* three groups with 0.2 proportion of DEGs
* two batches

```{r group and batch, collapse=TRUE}
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(batchCells = c(500, 500),
                     nGenes = 5000,
                     de.prob = 0.2,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$batch)
table(cell_info$group)
## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result)
### fc
fc_group2_to_group3 <- gene_info$DEFacGroup3/gene_info$DEFacGroup2
```


## Simulate with five different modes

SCRIP contains five different simulation modes, and you can specify which mode do you use (default is GP-trendedBCV):

1. GP-trendedBCV
2. GP-commonBCV
3. BGP-commonBCV
4. BP
5. BGP-trendedBCV

```{r}
## BGP-commonBCV
simulate_result <- simmethods::SCRIP_simulation(
  ref_data = ref_data,
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(batchCells = 500,
                     nGenes = 1000,
                     mode = "BGP-commonBCV"),
  seed = 111
)
```
