#' Estimate Parameters From Real Datasets by SplatPop
#'
#' This function is used to estimate useful parameters from a real dataset by
#' using \code{splatPopEstimate} function in Splatter package.
#'
#' @param ref_data A count matrix. Each row represents a gene and each column
#' represents a cell.
#' @param verbose Logical.
#' @param seed An integer of a random seed.
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs and other variables are usually customed, so before simulating a dataset
#' you must point it out. See `Details` below for more information.
#' @importFrom splatter splatPopEstimate
#'
#' @return A list contains the estimated parameters and the results of execution
#' detection.
#' @export
#' @details
#' When estimating parameters with SplatPop, users can input some extra information
#' (but it is rarely used):
#' 1. means. Matrix of real gene means across a population, where each row is a gene and each column is an individual in the population.
#' 2. eqtl. A data.frame with all or top eQTL pairs from a real eQTL analysis. Must include columns: 'gene_id', 'pval_nominal', and 'slope'.
#' For more information, please check [splatter::splatPopEstimate] or <https://www.bioconductor.org/packages/release/bioc/vignettes/splatter/inst/doc/splatPop.html>
#' @references
#' Zappia, L., Phipson, B. & Oshlack, A. Splatter: simulation of single-cell RNA sequencing data. Genome Biology 18, 174 (2017). <https://doi.org/10.1186/s13059-017-1305-0>
#'
#' Bioconductor URL: <https://bioconductor.org/packages/release/bioc/html/splatter.html>
#'
#' Github URL: <https://github.com/Oshlack/splatter>
#' @examples
#' ref_data <- simmethods::data
#' # Estimate parameters
#' estimate_result <- SplatPop_estimation(ref_data = ref_data,
#'                                        verbose = TRUE,
#'                                        seed = 111)
#' estimate_result <- estimate_result[["estimate_result"]]
#' ## Check the class
#' class(estimate_result) == "SplatPopParams"
SplatPop_estimation <- function(ref_data,
                                verbose = FALSE,
                                other_prior = NULL,
                                seed){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!is.matrix(ref_data)){
    ref_data <- as.matrix(ref_data)
  }
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    cat("Estimating parameters using SplatPop\n")
  }
  # Seed
  set.seed(seed)
  # Estimation
  tryCatch({
    # Estimate parameters from real data and return parameters and detection results
    estimate_detection <- peakRAM::peakRAM(
      estimate_result <- splatter::splatPopEstimate(ref_data,
                                                    means = other_prior[["means"]],
                                                    eqtl = other_prior[["eqtl"]])
      )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  estimate_output <- list(estimate_result = estimate_result,
                          estimate_detection = estimate_detection)
  return(estimate_output)
}


#' Simulate Datasets by SplatPop
#'
#' @param parameters A object generated by [splatter::splatPopEstimate()]
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs and other variables are usually customed, so before simulating a dataset
#' you must point it out.
#' @param return_format A character. Alternatives choices: list, SingleCellExperiment,
#' Seurat, h5ad. If you select `h5ad`, you will get a path where the .h5ad file saves to.
#' @param verbose Logical. Whether to return messages or not.
#' @param seed A random seed.
#' @importFrom splatter splatPopSimulate mockVCF
#' @details
#' In addtion to simulate datasets with default parameters, users want to simulate
#' other kinds of datasets, e.g. a counts matrix with 2 or more cell groups. In
#' SplatPop, you can set extra parameters to simulate datasets.
#'
#' The customed parameters you can set are below:
#' 1. nCells. In Splat, you can not set nCells directly and should set batchCells instead. For example, if you want to simulate 1000 cells, you can type `other_prior = list(batchCells = 1000)`. If you type `other_prior = list(batchCells = c(500, 500))`, the simulated data will have two batches.
#' 2. nGenes. You can directly set `other_prior = list(nGenes = 5000)` to simulate 5000 genes.
#' 3. nGroups. You can not directly set `other_prior = list(nGroups = 3)` to simulate 3 groups. Instead, you should set `other_prior = list(prob.group = c(0.2, 0.3, 0.5))` where the sum of group probabilities must equal to 1.
#' 4. de.prob. You can directly set `other_prior = list(de.prob = 0.2)` to simulate DEGs that account for 20 percent of all genes.
#' 5. prob.group. You can directly set `other_prior = list(prob.group = c(0.2, 0.3, 0.5))` to assign three proportions of cell groups. Note that the number of groups always equals to the length of the vector.
#' 6. nBatches. You can not directly set `other_prior = list(nBatches = 3)` to simulate 3 batches. Instead, you should set `other_prior = list(batchCells = c(500, 500, 500))` to reach the goal and the total cells are 1500.
#'
#' For more customed parameters in SplatPop, please check [splatter::SplatPopParams()].
#'
#' @export
#' @references
#' Zappia, L., Phipson, B. & Oshlack, A. Splatter: simulation of single-cell RNA sequencing data. Genome Biology 18, 174 (2017). <https://doi.org/10.1186/s13059-017-1305-0>
#'
#' Bioconductor URL: <https://bioconductor.org/packages/release/bioc/html/splatter.html>
#'
#' Github URL: <https://github.com/Oshlack/splatter>
SplatPop_simulation <- function(parameters,
                                other_prior,
                                return_format,
                                verbose = FALSE,
                                seed
){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  assertthat::assert_that(class(parameters) == "SplatPopParams")
  if(!is.null(other_prior)){
    parameters <- simutils::set_parameters(parameters = parameters,
                                           other_prior = other_prior,
                                           method = "SplatPop")
  }
  if(!is.null(other_prior[["prob.group"]])){
    parameters <- splatter::setParam(parameters,
                                     name = "group.prob",
                                     value = other_prior[["prob.group"]])

  }
  # Get params to check
  params_check <- splatter::getParams(parameters, c("nBatches",
                                                    "batchCells",
                                                    "nGroups",
                                                    "group.prob",
                                                    "de.prob",
                                                    "nCells",
                                                    "nGenes"))

  # check batches
  if(length(params_check[["batchCells"]]) > 1){
    assertthat::assert_that(params_check[["nBatches"]] > 1)
  }else{
    assertthat::assert_that(params_check[["nBatches"]] == 1)
  }
  # check groups
  if(length(params_check[["group.prob"]]) > 1){
    assertthat::assert_that(params_check[["nGroups"]] > 1)
  }else{
    assertthat::assert_that(params_check[["nGroups"]] == 1)
  }
  # DEGs proportion
  de.prob <- params_check[["de.prob"]]
  # Return to users
  cat(glue::glue("nCells: {params_check[['nCells']]}"), "\n")
  cat(glue::glue("nGenes: {params_check[['nGenes']]}"), "\n")
  cat(glue::glue("nGroups: {params_check[['nGroups']]}"), "\n")
  cat(glue::glue("de.prob: {de.prob}"), "\n")
  cat(glue::glue("nBatches: {params_check[['nBatches']]}"), "\n")
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    cat("Simulating datasets using SplatPop\n")
  }
  # Seed
  parameters <- splatter::setParam(parameters, name = "seed", value = seed)
  # vcf
  if(is.null(other_prior[["vcf"]])){
    vcf <- splatter::mockVCF(n.samples = 1,
                             seed = seed)
  }else vcf <- other_prior[["vcf"]]
  # gff
  if(is.null(other_prior[["gff"]])){
    gff <- splatter::mockGFF(n.genes = params_check[['nGenes']],
                             seed = seed)
  }else gff <- other_prior[["gff"]]
  # Estimation
  tryCatch({
    if(!is.null(other_prior[["paths"]])){
      cat("Simulating trajectory datasets by SplatPop")
      submethod <- "paths"
    }else{
      if(params_check[["nGroups"]] == 1){
        submethod <- "single"
      }else if(params_check[["nGroups"]] != 1){
        submethod <- "groups"
      }
    }
    simulate_detection <- peakRAM::peakRAM(
      simulate_result <- splatter::splatPopSimulate(parameters,
                                                    method = submethod,
                                                    vcf = vcf,
                                                    gff = gff,
                                                    eqtl = other_prior[["eqtl"]],
                                                    means = other_prior[["means"]],
                                                    key = other_prior[["key"]],
                                                    verbose = verbose)
    )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  # counts
  counts <- SingleCellExperiment::counts(simulate_result)
  colnames(counts) <- paste0("Cell", 1:ncol(counts))
  rownames(counts) <- paste0("Gene", as.numeric(stringr::str_split(rownames(counts), "_", simplify = TRUE)[, 2]))
  # col_data
  col_data <- as.data.frame(SummarizedExperiment::colData(simulate_result))
  col_data <- col_data[, c("Cell", "Batch", "Group")]
  colnames(col_data) <- c("cell_name", "batch", "group")
  col_data$cell_name <- paste0("Cell", 1:ncol(counts))
  rownames(col_data) <- col_data$cell_name
  # row_data
  row_data <- as.data.frame(SummarizedExperiment::rowData(simulate_result))

  if(params_check[['nGroups']] == 1){
    row_data <- data.frame("gene_name" = paste0("Gene", 1:nrow(counts)))
  }else{
    group_fac <- row_data[, grep(colnames(row_data), pattern = "^GroupDE")]
    total_sum <- rowSums(group_fac)
    de_gene <- ifelse(total_sum == params_check[['nGroups']], "no", "yes")
    row_data <- data.frame("gene_name" = paste0("Gene", 1:nrow(counts)),
                           "de_gene" = de_gene)
    row_data <- cbind(row_data, group_fac)
    colnames(row_data) <- c("gene_name", "de_gene", paste0("DEFacGroup", 1:params_check[['nGroups']]))
    rownames(row_data) <- row_data$gene_name
  }
  simulate_result <- simutils::data_conversion(SCE_object = simulate_result,
                                               return_format = return_format)

  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  simulate_output <- list(simulate_result = simulate_result,
                          simulate_detection = simulate_detection)
  return(simulate_output)
}

