#' Estimate Parameters From Real Datasets by Lun2
#'
#' This function is used to estimate useful parameters from a real dataset by
#' using \code{lun2Estimate} function in Splatter package.
#'
#' @param ref_data A count matrix. Each row represents a gene and each column
#' represents a cell.
#' @param verbose Logical.
#' @param seed An integer of a random seed.
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @importFrom peakRAM peakRAM
#' @importFrom splatter lun2Estimate
#'
#' @return A list contains the estimated parameters and the results of execution
#' detection.
#' @export
#' @details
#' When you use Lun2 to estimate parameters from a real dataset, you must input
#' a numeric vector to specify the groups or plates that each cell comes from,
#' like `other_prior = list(group.condition = the numeric vector)`. See `Examples`
#' and learn from it.
#' @references
#' Zappia L, Phipson B, Oshlack A. Splatter: simulation of single-cell RNA sequencing data[J]. Genome biology, 2017, 18(1): 1-15. https://doi.org/10.1186/s13059-017-1305-0
#'
#' Bioconductor URL: <https://bioconductor.org/packages/release/bioc/html/splatter.html>
#'
#' Github URL: <https://github.com/Oshlack/splatter>
#' @examples
#' ref_data <- simmethods::data
#' group_condition <- simmethods::group_condition
#' ## group_condition can must be a numeric vector.
#' other_prior <- list(group.condition = as.numeric(group_condition))
#' ## Estimation
#' estimate_result <- simmethods::Lun2_estimation(ref_data = ref_data,
#'                                                other_prior = other_prior,
#'                                                verbose = TRUE,
#'                                                seed = 10)
#' estimate_result <- estimate_result[["estimate_result"]]
## Check the class
#' class(estimate_result) == "Lun2Params"
Lun2_estimation <- function(ref_data,
                            verbose = FALSE,
                            other_prior = NULL,
                            seed){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!is.matrix(ref_data)){
    ref_data <- as.matrix(ref_data)
  }
  if(is.null(other_prior[["group.condition"]])){
    stop("Please input the information of group condition")
  }else{
    other_prior[["plates"]] <- other_prior[["group.condition"]]
  }
  if(is.null(other_prior[["min.size"]])){
    min.size <- 50
  }
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    cat("Estimating parameters using Lun2\n")
  }
  # Seed
  set.seed(seed)
  # Estimation
  tryCatch({
    # Estimate parameters from real data and return parameters and detection results
    estimate_detection <- peakRAM::peakRAM(
      estimate_result <- splatter::lun2Estimate(ref_data,
                                                plates = other_prior[["plates"]],
                                                min.size = min.size,
                                                verbose = verbose)
    )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  estimate_output <- list(estimate_result = estimate_result,
                          estimate_detection = estimate_detection)
  return(estimate_output)
}


#' Simulate Datasets by Lun2
#'
#' @param parameters A object generated by [splatter::lun2Estimate()]
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @param return_format A character. Alternatives choices: list, SingleCellExperiment,
#' Seurat, h5ad. If you select `h5ad`, you will get a path where the .h5ad file saves to.
#' @param verbose Logical. Whether to return messages or not.
#' @param seed A random seed.
#' @importFrom splatter lun2Simulate
#' @export
#' @details
#' In addtion to simulate datasets with default parameters, users want to simulate
#' other kinds of datasets, e.g. a counts matrix with 2 or more cell groups. In
#' Lun2, you can set extra parameters to simulate datasets.
#'
#' The customed parameters you can set are below:
#' 1. nCells. In Lun2, you can not set nCells directly and should set cell.plates instead. For example, if you want to simulate 1000 cells, you can type `other_prior = list(cell.plates = sample(1:3, 1000, replace = TRUE))` and you will get three plates or groups of cells. If you only want to simulate one plate or group, just type `other_prior = list(cell.plates = rep(1, 1000))`
#' 2. nGenes. You can directly set `other_prior = list(nGenes = 5000)` to simulate 5000 genes.
#' 3. de.prob. You can directly set `other_prior = list(de.prob = 0.2)` to simulate DEGs that account for 20 percent of all genes.
#' 4. fc.group. You can directly set `other_prior = list(fc.group. = 2)` to specify the foldchange of DEGs.
#'
#' For more customed parameters in Lun2, please check [splatter::Lun2Params()].
#' @references
#' Zappia L, Phipson B, Oshlack A. Splatter: simulation of single-cell RNA sequencing data[J]. Genome biology, 2017, 18(1): 1-15. <https://doi.org/10.1186/s13059-017-1305-0>
#'
#' Bioconductor URL: <https://bioconductor.org/packages/release/bioc/html/splatter.html>
#'
#' Github URL: <https://github.com/Oshlack/splatter>
#'
#' @examples
#' ref_data <- simmethods::data
#' group_condition <- simmethods::group_condition
#' ## group_condition must be a numeric vector
#' other_prior <- list(group.condition = as.numeric(group_condition))
#' ## Estimation
#' estimate_result <- simmethods::Lun2_estimation(ref_data = ref_data,
#'                                                other_prior = other_prior,
#'                                                verbose = TRUE,
#'                                                seed = 10)
#'
#'
#' # (1) Simulate 500 cells (Since we can not set nCells directly, so we can set
#' # cell.plates (a numeric vector)) and 2000 genes
#' simulate_result <- simmethods::Lun2_simulation(parameters = estimate_result[["estimate_result"]],
#'                                                other_prior = list(cell.plates = sample(1:2, 500, replace = TRUE),
#'                                                                   nGenes = 2000),
#'                                                return_format = "list",
#'                                                verbose = TRUE,
#'                                                seed = 111)
#' count_data <- simulate_result[["simulate_result"]][["count_data"]]
#' dim(count_data)
#' ## col_data
#' col_data <- simulate_result[["simulate_result"]][["col_meta"]]
#' head(col_data)
#' ## row_data
#' row_data <- simulate_result[["simulate_result"]][["row_meta"]]
#' head(row_data)
#'
#'
#' # (2) Simulate two groups (de.prob = 0.2, fc.group = 2)
#' simulate_result <- simmethods::Lun2_simulation(parameters = estimate_result[["estimate_result"]],
#'                                                other_prior = list(cell.plates = sample(1:2, 500, replace = TRUE),
#'                                                                   nGenes = 10000,
#'                                                                   de.prob = 0.2,
#'                                                                   fc.group = 2),
#'                                                return_format = "list",
#'                                                verbose = TRUE,
#'                                                seed = 111)
#' count_data <- simulate_result[["simulate_result"]][["count_data"]]
#' dim(count_data)
#' ## cell information
#' col_data <- simulate_result[["simulate_result"]][["col_meta"]]
#' head(col_data)
#' table(col_data$plate)
#' ## gene information
#' row_data <- simulate_result[["simulate_result"]][["row_meta"]]
#' head(row_data)
#' ### The result of Lun2 contains the factors of different groups and uses can
#' ### calculate the fold change by division. For example, the DEFactors of A gene
#' ### in Ingroup and Outgroup are respectively 2 and 1, and the fold change of A gene
#' ### is 2/1=2 or 1/2=0.5.
#' fc_ingroup_to_outgroup <- row_data$DEFacIngroup/row_data$DEFacOutgroup
#' table(fc_ingroup_to_outgroup != 1)[2]/10000 ## de.prob = 0.2
#' ### number of all DEGs
#' table(row_data$de_gene)[2]/10000 ## de.prob = 0.2
Lun2_simulation <- function(parameters,
                            other_prior = NULL,
                            return_format,
                            verbose = FALSE,
                            seed
){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  assertthat::assert_that(class(parameters) == "Lun2Params")

  if(!is.null(other_prior)){
    parameters <- simutils::set_parameters(parameters = parameters,
                                           other_prior = other_prior,
                                           method = "Lun2")
  }
  # de.prob
  if(!is.null(other_prior[["de.prob"]])){
    nPlates <- splatter::getParam(parameters, "nPlates")
    parameters <- splatter::setParam(parameters,
                                     name = "de.nGenes",
                                     value = round(other_prior[["de.prob"]]*parameters@nGenes))
    de.prob <- other_prior[["de.prob"]]
  }else{
    de.prob <- parameters@de.nGenes/parameters@nGenes
  }
  # fc.group
  if(!is.null(other_prior[["fc.group"]])){
    parameters <- splatter::setParam(parameters,
                                     name = "de.fc",
                                     value = other_prior[["fc.group"]])
  }
  # Get params to check
  params_check <- splatter::getParams(parameters, c("nCells",
                                                    "nGenes",
                                                    "nPlates",
                                                    "de.nGenes",
                                                    "de.fc"))

  # Return to users
  cat(glue::glue("nCells: {params_check[['nCells']]}"), "\n")
  cat(glue::glue("nGenes: {params_check[['nGenes']]}"), "\n")
  cat(glue::glue("nPlates: {params_check[['nPlates']]}"), "\n")
  cat(glue::glue("de.prob: {de.prob}"), "\n")
  cat(glue::glue("fc.group: {params_check[['de.fc']]}"), "\n")
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    cat("Simulating datasets using Lun2\n")
  }
  # Seed
  parameters <- splatter::setParam(parameters, name = "seed", value = seed)
  # zinb
  if(is.null(other_prior[["zinb"]])){
    zinb <- FALSE
  }else zinb <- other_prior[["zinb"]]
  # Estimation
  tryCatch({
    simulate_detection <- peakRAM::peakRAM(
      simulate_result <- splatter::lun2Simulate(parameters,
                                                verbose = verbose,
                                                zinb = zinb)
    )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  # counts
  counts <- as.matrix(SingleCellExperiment::counts(simulate_result))
  # col_data
  col_data <- as.data.frame(SummarizedExperiment::colData(simulate_result))
  if(de.prob == 0){
    col_data <- col_data[, c("Cell", "Plate")]
    colnames(col_data) <- c("cell_name", "plate")
  }else{
    col_data <- col_data[, c("Cell", "Plate", "Ingroup")]
    colnames(col_data) <- c("cell_name", "plate", "ingroup")
  }

  # row_data
  if(de.prob == 0){
    row_data <- data.frame("gene_name" = paste0("Gene", 1:nrow(counts)))
    rownames(row_data) <- row_data$gene_name
  }else{
    row_data <- as.data.frame(SummarizedExperiment::rowData(simulate_result))
    group_fac <- row_data[, grep(colnames(row_data), pattern = "^DEFac")]
    total_sum <- rowSums(group_fac)
    de_gene <- ifelse(total_sum == 2, "no", "yes")
    row_data[, 2] <- de_gene
    row_data <- row_data[, 1:2]
    row_data <- cbind(row_data, group_fac)
    colnames(row_data) <- c("gene_name", "de_gene", colnames(group_fac))
  }

  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- simutils::data_conversion(SCE_object = simulate_result,
                                               return_format = return_format)

  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  simulate_output <- list(simulate_result = simulate_result,
                          simulate_detection = simulate_detection)
  return(simulate_output)
}

