#' Estimate Parameters From Real Datasets by ESCO
#'
#' This function is used to estimate useful parameters from a real dataset by
#' using \code{escoEstimate} function in ESCO package.
#'
#' @param ref_data A count matrix. Each row represents a gene and each column
#' represents a cell.
#' @param verbose Logical.
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @param seed An integer of a random seed.
#' @importFrom ESCO escoEstimate newescoParams
#' @return A list contains the estimated parameters and the results of execution
#' detection.
#' @export
#'
#' @references
#' Tian J, Wang J, Roeder K. ESCO: single cell expression simulation incorporating gene co-expression. Bioinformatics, 2021, 37(16): 2374-2381. <https://doi.org/10.1093/bioinformatics/btab116>
#'
#' Github URL: <https://github.com/JINJINT/ESCO>
ESCO_estimation <- function(ref_data,
                            other_prior = NULL,
                            verbose = FALSE,
                            seed){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("ESCO", quietly = TRUE)){
    cat("ESCO is not installed on your device\n")
    cat("Installing ESCO...\n")
    devtools::install_github("JINJINT/ESCO")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!is.matrix(ref_data)){
    ref_data <- as.matrix(ref_data)
  }
  other_prior[["counts"]] <- ref_data
  if(is.null(other_prior[["group.condition"]])){
    other_prior[["group"]] <- FALSE
    other_prior[["cellinfo"]] <- NULL
  }else{
    other_prior[["group"]] <- TRUE
    other_prior[["cellinfo"]] <- other_prior[["group.condition"]]
  }
  other_prior[["params"]] <- ESCO::newescoParams()
  other_prior[["dirname"]] <- tempdir()
  ## tree
  if(!is.null(other_prior[["tree"]]) | !is.null(other_prior[["paths"]])){
    tree <- simutils::make_trees(ref_data = ref_data,
                                 group = other_prior[["cellinfo"]],
                                 is_Newick = FALSE,
                                 is_parenthetic = TRUE,
                                 return_group = TRUE)
    group <- tree[["group"]]
    tree <- tree[["phyla"]]
  }else{
    tree <- NULL
  }
  estimate_formals <- simutils::change_parameters(function_expr = "ESCO::escoEstimate",
                                                  other_prior = other_prior,
                                                  step = "estimation")
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    cat("Estimating parameters using ESCO\n")
  }
  # Seed
  set.seed(seed)
  # Estimation
  tryCatch({
    # Estimate parameters from real data and return parameters and detection results
    estimate_detection <- peakRAM::peakRAM(
      estimate_result <- do.call(ESCO::escoEstimate, estimate_formals)
    )
  }, error = function(e){
    as.character(e)
  })
  if(!is.null(tree)){
    estimate_result <- list(estimate_result = estimate_result,
                            tree = tree,
                            group = group)
  }
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  estimate_output <- list(estimate_result = estimate_result,
                          estimate_detection = estimate_detection)
  return(estimate_output)
}


#' Simulate Datasets by ESCO
#'
#' @param parameters A object generated by [ESCO::escoEstimate()]
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @param return_format A character. Alternative choices: list, SingleCellExperiment,
#' Seurat, h5ad. If you select `h5ad`, you will get a path where the .h5ad file saves to.
#' @param verbose Logical. Whether to return messages or not.
#' @param seed A random seed.
#' @importFrom ESCO escoEstimate
#' @export
#' @references
#' Tian J, Wang J, Roeder K. ESCO: single cell expression simulation incorporating gene co-expression. Bioinformatics, 2021, 37(16): 2374-2381. <https://doi.org/10.1093/bioinformatics/btab116>
#'
#' Github URL: <https://github.com/JINJINT/ESCO>
#'
ESCO_simulation <- function(parameters,
                            return_format,
                            other_prior = NULL,
                            verbose = FALSE,
                            seed
){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("ESCO", quietly = TRUE)){
    cat("ESCO is not installed on your device\n")
    cat("Installing ESCO...\n")
    devtools::install_github("JINJINT/ESCO")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(length(parameters) == 3){
    tree <- parameters[["tree"]]
    group <- parameters[["group"]]
    parameters <- parameters[["estimate_result"]]
  }else{
    group <- NULL
    tree <- NULL
  }
  assertthat::assert_that(class(parameters) == "escoParams")
  if(!is.null(other_prior)){
    parameters <- simutils::set_parameters(parameters = parameters,
                                           other_prior = other_prior,
                                           method = "ESCO")
  }
  # nCells
  if(!is.null(other_prior[["nCells"]])){
    parameters <- splatter::setParam(parameters, name = "nCells", value = other_prior[["nCells"]])
  }
  # nGenes
  if(!is.null(other_prior[["nGenes"]])){
    parameters <- splatter::setParam(parameters, name = "nGenes", value = other_prior[["nGenes"]])
  }
  # nGroup
  if(!is.null(group)){
    parameters <- splatter::setParam(parameters, name = "nGroups", value = length(unique(group)))
  }
  if(!is.null(other_prior[["prob.group"]])){
    parameters <- splatter::setParam(parameters,
                                     name = "group.prob",
                                     value = other_prior[["prob.group"]])

  }else{
    if(!is.null(tree)){
      nGroups <- splatter::getParam(parameters, name = "nGroups")
      prob.group <- c(rep(1/nGroups, nGroups-1),
                      1 - c(1/nGroups * c(nGroups-1)))
      parameters <- splatter::setParam(parameters,
                                       name = "group.prob",
                                       value = prob.group)
    }
  }
  # Get params to check
  params_check <- splatter::getParams(parameters, c("nCells",
                                                    "nGenes",
                                                    "nGroups",
                                                    "group.prob",
                                                    "de.prob"))

  # Return to users
  cat(glue::glue("nCells: {params_check[['nCells']]}"), "\n")
  cat(glue::glue("nGenes: {params_check[['nGenes']]}"), "\n")
  cat(glue::glue("nGroups: {params_check[['nGroups']]}"), "\n")
  cat(glue::glue("de.group: {params_check[['de.prob']]}"), "\n")
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    cat("Simulating datasets using ESCO\n")
  }
  # Seed
  parameters <- splatter::setParam(parameters, name = "seed", value = seed)
  parameters <- splatter::setParam(parameters,
                                   name = "deall.prob",
                                   value = params_check[['de.prob']])
  # Estimation
  tryCatch({
    if(!is.null(other_prior[["tree"]])){
      cat("Simulating trajectory of trees datasets by ESCO \n")
      submethod <- "tree"
      parameters <- splatter::setParam(parameters, name = "tree", value = tree)
    }else if(!is.null(other_prior[["paths"]])){
      cat("Simulating trajectory datasets by ESCO")
      submethod <- "traj"
      parameters <- splatter::setParam(parameters, name = "tree", value = tree)
    }else{
      if(params_check[["nGroups"]] == 1){
        submethod <- "single"
      }else if(params_check[["nGroups"]] != 1){
        submethod <- "group"
      }
    }
    simulate_detection <- peakRAM::peakRAM(
      simulate_result <- ESCO::escoSimulate(parameters,
                                            type = submethod,
                                            verbose = verbose))
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  # counts
  counts <- as.matrix(SingleCellExperiment::counts(simulate_result))
  # col_data
  col_data <- as.data.frame(SummarizedExperiment::colData(simulate_result))
  if(params_check[['nGroups']] == 1){
    col_data <- data.frame("cell_name" = colnames(counts),
                           "group" = rep("Group1", ncol(counts)))
  }else{
    col_data <- data.frame("cell_name" = colnames(counts),
                           "group" = col_data$Group)
  }
  rownames(col_data) <- col_data$cell_name
  # row_data
  row_data <- as.data.frame(SummarizedExperiment::rowData(simulate_result))
  if(params_check[['nGroups']] == 1){
    row_data <- data.frame("gene_name" = rownames(counts))
    rownames(row_data) <- row_data$gene_name
  }else{
    group_fac <- row_data[, grep(colnames(row_data), pattern = "^DEFac")]
    total_sum <- rowSums(group_fac)
    de_gene <- ifelse(total_sum == params_check[['nGroups']], "no", "yes")
    row_data[, 2] <- de_gene
    row_data <- row_data[, -c(3:5, ncol(row_data))]
    colnames(row_data) <- c("gene_name", "de_gene", colnames(group_fac))
  }
  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- simutils::data_conversion(SCE_object = simulate_result,
                                               return_format = return_format)

  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  simulate_output <- list(simulate_result = simulate_result,
                          simulate_detection = simulate_detection)
  return(simulate_output)
}

