#' Estimate Parameters From Real Datasets by SparseDC
#'
#' This function is used to estimate useful parameters from a real dataset by
#' using \code{sparseDCEstimate} function in Splatter package.
#'
#' @param ref_data A count matrix. Each row represents a gene and each column
#' represents a cell.
#' @param verbose Logical.
#' @param seed An integer of a random seed.
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @importFrom splatter sparseDCEstimate newSparseDCParams
#' @return A list contains the estimated parameters and the results of execution
#' detection.
#' @export
#' @references
#' Barron M, Zhang S, Li J. A sparse differential clustering algorithm for tracing cell type changes via single-cell RNA-sequencing data. Nucleic acids research, 2018, 46(3): e14-e14. <https://doi.org/10.1093/nar/gkx1113>
#'
#' CRAN URL: <https://cran.rstudio.com/web/packages/SparseDC/index.html>
#'
SparseDC_estimation <- function(ref_data,
                                verbose = FALSE,
                                other_prior,
                                seed){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!is.matrix(ref_data)){
    ref_data <- as.matrix(ref_data)
  }
  other_prior[["counts"]] <- ref_data
  if(is.null(other_prior[["group.condition"]])){
    stop("Please input the conditions that each cell belongs to")
  }
  other_prior[["conditions"]] <- other_prior[["group.condition"]]
  if(is.null(other_prior[["nclusters"]])){
    other_prior[["nclusters"]] <- 2
  }
  estimate_formals <- simutils::change_parameters(function_expr = "splatter::sparseDCEstimate",
                                                  other_prior = other_prior,
                                                  step = "estimation")
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    cat("Estimating parameters using SparseDC\n")
  }
  # Seed
  set.seed(seed)
  # Estimation
  tryCatch({
    # Estimate parameters from real data and return parameters and detection results
    estimate_detection <- peakRAM::peakRAM(
      estimate_result <- splatter::sparseDCEstimate(counts = estimate_formals[["counts"]],
                                                    conditions = estimate_formals[["conditions"]],
                                                    nclusters = estimate_formals[["nclusters"]],
                                                    norm = estimate_formals[["norm"]],
                                                    params = splatter::newSparseDCParams())
    )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  estimate_output <- list(estimate_result = estimate_result,
                          estimate_detection = estimate_detection)
  return(estimate_output)
}


#' Simulate Datasets by SparseDC
#'
#' @param parameters A object generated by [splatter::sparseDCEstimate()]
#' @param other_prior A list with names of certain parameters. Some methods need
#' extra parameters to execute the estimation step, so you must input them. In
#' simulation step, the number of cells, genes, groups, batches, the percent of
#' DEGs are usually customed, so before simulating a dataset you must point it out.
#' See `Details` below for more information.
#' @param return_format A character. Alternative choices: list, SingleCellExperiment,
#' Seurat, h5ad. If you select `h5ad`, you will get a path where the .h5ad file saves to.
#' @param verbose Logical. Whether to return messages or not.
#' @param seed A random seed.
#' @importFrom splatter getParams setParam sparseDCSimulate
#' @importFrom stringr str_replace
#' @export
#' @references
#' Barron M, Zhang S, Li J. A sparse differential clustering algorithm for tracing cell type changes via single-cell RNA-sequencing data. Nucleic acids research, 2018, 46(3): e14-e14. <https://doi.org/10.1093/nar/gkx1113>
#'
#' CRAN URL: <https://cran.rstudio.com/web/packages/SparseDC/index.html>
SparseDC_simulation <- function(parameters,
                                other_prior = NULL,
                                return_format,
                                verbose = FALSE,
                                seed
){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("splatter", quietly = TRUE)){
    cat("Splatter is not installed on your device\n")
    cat("Installing splatter...\n")
    BiocManager::install("splatter")
  }
  ## nCells
  if(!is.null(other_prior[["nCells"]])){
    parameters <- splatter::setParam(parameters, "nCells", other_prior[["nCells"]])
  }
  ## nGenes
  if(!is.null(other_prior[["nGenes"]])){
    parameters <- splatter::setParam(parameters, "nGenes", other_prior[["nGenes"]])
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  assertthat::assert_that(class(parameters) == "SparseDCParams")

  # Get params to check
  params_check <- splatter::getParams(parameters, c("nCells",
                                                    "nGenes"))
  # Return to users
  cat(glue::glue("nCells: {params_check[['nCells']]}"), "\n")
  cat(glue::glue("nGenes: {params_check[['nGenes']]}"), "\n")
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    cat("Simulating datasets using SparseDC\n")
  }
  # Seed
  parameters <- splatter::setParam(parameters, name = "seed", value = seed)
  # Estimation
  tryCatch({
    simulate_detection <- peakRAM::peakRAM(
      simulate_result <- splatter::sparseDCSimulate(parameters, verbose = verbose)
    )
  }, error = function(e){
    as.character(e)
  })
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  simulate_result <- simutils::data_conversion(SCE_object = simulate_result,
                                               return_format = return_format)

  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  simulate_output <- list(simulate_result = simulate_result,
                          simulate_detection = simulate_detection)
  return(simulate_output)
}

