#' Get Information of PROSSTT
#'
#' @param ... ...
#'
#' @return A list contains the information of method and default parameters
#' @import simutils
#' @export
#'
#' @examples
#' PROSSTT_method_definition <- PROSSTT_method_definition()
PROSSTT_method_definition <- function(...){

  PROSSTT_parameters <- parameter_sets(
    param_reference(
      id = "input_refernece",
      type = c("matrix"),
      default = NULL,
      process = "estimation",
      description = "Note that the input matrix for estimation is nessesary unless you can provide newick tree format string."
    ),
    param_others(
      id = "newick_tree_format_string",
      type = "string",
      default = "(B:100,C:60)A:70;",
      description = "Usually it is learned by simutils::make_tree function. Note that the number in the string refers to the pseudotime which is similar to the length"
    ),
    param_integer(
      id = "modules",
      default = 15L,
      lower = 1L,
      description = "Total number of expression programs for the lineage tree."
    ),
    param_integer(
      id = "nGenes",
      default = 10000L,
      lower = 1L,
      description = "Total number of genes."
    ),
    param_integer(
      id = "nCells",
      default = 100L,
      lower = 1L
    ),
    param_numeric(
      id = "rel_exp_cutoff",
      default = 8,
      lower = 0,
      border = FALSE,
      upper = 9,
      description = "The log threshold for the maximum average expression before scaling."
    ),
    param_numeric(
      id = "intra_branch_tol",
      default = 0.5,
      lower = 0,
      border = FALSE,
      upper = 1,
      description = "The threshold for correlation between expression programs in the same branch."
    ),
    param_numeric(
      id = "inter_branch_tol",
      default = 0,
      lower = 0,
      border = TRUE,
      upper = 1,
      description = "The threshold for anticorrelation between relative gene expression in parallel branches."
    ),
    param_numeric(
      id = "a",
      default = 0.5,
      lower = 0,
      border = FALSE,
      description = "Float a if coefficients are generated by a Gamma distribution."
    ),
    param_numeric(
      id = "alpha",
      default = 0.2,
      description = "Average alpha value."
    ),
    param_numeric(
      id = "beta",
      default = 3,
      description = "Average beta value."
    ),
    param_integer(
      id = "seed",
      default = 687680L
    ))

  PROSSTT_method <- method_definition(
    method = "PROSSTT",
    programming = "Python",
    url = "http://wwwuser.gwdg.de/~compbiol/prosstt/doc/",
    authors = authors_definition(
      first = "Nikolaos",
      last = "Papadopoulos",
      email = NULL,
      github = "https://github.com/soedinglab/prosstt/",
      orcid = NULL
    ),
    manuscript = manuscript_definition(
      title = "PROSSTT: probabilistic simulation of single-cell RNA-seq data for complex differentiation processes",
      doi = "10.1093/bioinformatics/btz078",
      journal = "Bioinformatics",
      date = "2019",
      peer_review = TRUE
    ),
    description = "PROSSTT is a package with code for the simulation of scRNAseq data for dynamic processes such as cell differentiation.")

  list(PROSSTT_method = PROSSTT_method,
       PROSSTT_parameters = PROSSTT_parameters)
}
